{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using TensorFlow backend.\n"
     ]
    }
   ],
   "source": [
    "from __future__ import print_function\n",
    "import keras\n",
    "from keras.models import Sequential, Model, load_model\n",
    "from keras import backend as K\n",
    "\n",
    "import tensorflow as tf\n",
    "\n",
    "import os\n",
    "import pandas as pd\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.cm as cm\n",
    "\n",
    "import aparent.visualization as vis\n",
    "\n",
    "from aparent.predictor import *\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2>Load APARENT model</h2>\n",
    "<br/>\n",
    "-- Load APARENT, which has been trained to predict the isoform abundance and cut profile of a proximal PAS given a fixed background distal PAS (trained on random 3' UTR APA MPRA data).<br/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING:tensorflow:From /home/johli/anaconda3/envs/aparent/lib/python3.6/site-packages/tensorflow/python/framework/op_def_library.py:263: colocate_with (from tensorflow.python.framework.ops) is deprecated and will be removed in a future version.\n",
      "Instructions for updating:\n",
      "Colocations handled automatically by placer.\n",
      "WARNING:tensorflow:From /home/johli/anaconda3/envs/aparent/lib/python3.6/site-packages/keras/backend/tensorflow_backend.py:3445: calling dropout (from tensorflow.python.ops.nn_ops) with keep_prob is deprecated and will be removed in a future version.\n",
      "Instructions for updating:\n",
      "Please use `rate` instead of `keep_prob`. Rate should be set to `rate = 1 - keep_prob`.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/johli/anaconda3/envs/aparent/lib/python3.6/site-packages/keras/engine/saving.py:292: UserWarning: No training configuration found in save file: the model was *not* compiled. Compile it manually.\n",
      "  warnings.warn('No training configuration found in save file: '\n"
     ]
    }
   ],
   "source": [
    "#Load base APARENT model\n",
    "\n",
    "save_dir = os.path.join(os.getcwd(), '../../saved_models')\n",
    "model_name = 'aparent_large_lessdropout_all_libs_no_sampleweights.h5'\n",
    "model_path = os.path.join(save_dir, model_name)\n",
    "\n",
    "aparent_model = load_model(model_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "len(polyadb_df) = 294\n"
     ]
    }
   ],
   "source": [
    "#Load APADB Data and filter on targeted genes\n",
    "\n",
    "genes = ['RUNX1', 'CEBPA', 'GATA2', 'ANKRD26', 'DDX41', 'ETV6', 'PTEN', 'BRCA1', 'BRCA2', 'TP53', 'APC', 'ATM', 'PALB2', 'MSH2', 'MLH1', 'MSH6', 'PMS2', 'MUTYH']\n",
    "\n",
    "polyadb_df = pd.read_csv('polyadb_processed.csv', sep=',')\n",
    "\n",
    "polyadb_df = polyadb_df.loc[polyadb_df['gene'].isin(genes)].reset_index(drop=True).copy()\n",
    "\n",
    "print('len(polyadb_df) = ' + str(len(polyadb_df)))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "polyadb_df_minus = polyadb_df.query(\"strand == '-'\").copy().reset_index(drop=True)\n",
    "polyadb_df_plus = polyadb_df.query(\"strand == '+'\").copy().reset_index(drop=True)\n",
    "\n",
    "polyadb_df_minus = polyadb_df_minus.sort_values(by='pas_pos', ascending=False).copy().reset_index(drop=True)\n",
    "polyadb_df_plus = polyadb_df_plus.sort_values(by='pas_pos', ascending=True).copy().reset_index(drop=True)\n",
    "\n",
    "new_gene_id_list_plus = []\n",
    "sitenum_list_plus = []\n",
    "gene_id_dict = {}\n",
    "for _, row in polyadb_df_plus.iterrows() :\n",
    "\n",
    "    gene = row['gene']\n",
    "\n",
    "    if gene not in gene_id_dict :\n",
    "        gene_id_dict[gene] = 0\n",
    "\n",
    "    gene_id_dict[gene] += 1\n",
    "\n",
    "    new_gene_id_list_plus.append(gene + \".\" + str(gene_id_dict[gene]))\n",
    "    sitenum_list_plus.append(gene_id_dict[gene])\n",
    "\n",
    "polyadb_df_plus['gene_id'] = new_gene_id_list_plus\n",
    "polyadb_df_plus['sitenum'] = sitenum_list_plus\n",
    "\n",
    "new_gene_id_list_minus = []\n",
    "sitenum_list_minus = []\n",
    "gene_id_dict = {}\n",
    "for _, row in polyadb_df_minus.iterrows() :\n",
    "\n",
    "    gene = row['gene']\n",
    "\n",
    "    if gene not in gene_id_dict :\n",
    "        gene_id_dict[gene] = 0\n",
    "\n",
    "    gene_id_dict[gene] += 1\n",
    "\n",
    "    new_gene_id_list_minus.append(gene + \".\" + str(gene_id_dict[gene]))\n",
    "    sitenum_list_minus.append(gene_id_dict[gene])\n",
    "\n",
    "polyadb_df_minus['gene_id'] = new_gene_id_list_minus\n",
    "polyadb_df_minus['sitenum'] = sitenum_list_minus\n",
    "\n",
    "polyadb_df = pd.concat([polyadb_df_plus, polyadb_df_minus])\n",
    "\n",
    "polyadb_df = polyadb_df.sort_values(by=['gene', 'sitenum'], ascending=True).reset_index(drop=True).copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Predicting variants for PAS = ANKRD26.1\n",
      "Predicting variants for PAS = ANKRD26.2\n",
      "Predicting variants for PAS = ANKRD26.3\n",
      "Predicting variants for PAS = ANKRD26.4\n",
      "Predicting variants for PAS = ANKRD26.5\n",
      "Predicting variants for PAS = ANKRD26.6\n",
      "Predicting variants for PAS = ANKRD26.7\n",
      "Predicting variants for PAS = ANKRD26.8\n",
      "Predicting variants for PAS = ANKRD26.9\n",
      "Predicting variants for PAS = ANKRD26.10\n",
      "Predicting variants for PAS = ANKRD26.11\n",
      "Predicting variants for PAS = ANKRD26.12\n",
      "Predicting variants for PAS = ANKRD26.13\n",
      "Predicting variants for PAS = ANKRD26.14\n",
      "Predicting variants for PAS = ANKRD26.15\n",
      "Predicting variants for PAS = ANKRD26.16\n",
      "Predicting variants for PAS = ANKRD26.17\n",
      "Predicting variants for PAS = ANKRD26.18\n",
      "Predicting variants for PAS = ANKRD26.19\n",
      "Predicting variants for PAS = ANKRD26.20\n",
      "Predicting variants for PAS = ANKRD26.21\n",
      "Predicting variants for PAS = ANKRD26.22\n",
      "Predicting variants for PAS = ANKRD26.23\n",
      "Predicting variants for PAS = ANKRD26.24\n",
      "Predicting variants for PAS = ANKRD26.25\n",
      "Predicting variants for PAS = ANKRD26.26\n",
      "Predicting variants for PAS = ANKRD26.27\n",
      "Predicting variants for PAS = ANKRD26.28\n",
      "Predicting variants for PAS = ANKRD26.29\n",
      "Predicting variants for PAS = APC.1\n",
      "Predicting variants for PAS = APC.2\n",
      "Predicting variants for PAS = APC.3\n",
      "Predicting variants for PAS = APC.4\n",
      "Predicting variants for PAS = APC.5\n",
      "Predicting variants for PAS = APC.6\n",
      "Predicting variants for PAS = APC.7\n",
      "Predicting variants for PAS = APC.8\n",
      "Predicting variants for PAS = APC.9\n",
      "Predicting variants for PAS = APC.10\n",
      "Predicting variants for PAS = APC.11\n",
      "Predicting variants for PAS = APC.12\n",
      "Predicting variants for PAS = APC.13\n",
      "Predicting variants for PAS = APC.14\n",
      "Predicting variants for PAS = APC.15\n",
      "Predicting variants for PAS = APC.16\n",
      "Predicting variants for PAS = APC.17\n",
      "Predicting variants for PAS = APC.18\n",
      "Predicting variants for PAS = ATM.1\n",
      "Predicting variants for PAS = ATM.2\n",
      "Predicting variants for PAS = ATM.3\n",
      "Predicting variants for PAS = ATM.4\n",
      "Predicting variants for PAS = ATM.5\n",
      "Predicting variants for PAS = ATM.6\n",
      "Predicting variants for PAS = ATM.7\n",
      "Predicting variants for PAS = ATM.8\n",
      "Predicting variants for PAS = ATM.9\n",
      "Predicting variants for PAS = ATM.10\n",
      "Predicting variants for PAS = ATM.11\n",
      "Predicting variants for PAS = ATM.12\n",
      "Predicting variants for PAS = ATM.13\n",
      "Predicting variants for PAS = ATM.14\n",
      "Predicting variants for PAS = ATM.15\n",
      "Predicting variants for PAS = ATM.16\n",
      "Predicting variants for PAS = ATM.17\n",
      "Predicting variants for PAS = BRCA1.1\n",
      "Predicting variants for PAS = BRCA1.2\n",
      "Predicting variants for PAS = BRCA1.3\n",
      "Predicting variants for PAS = BRCA1.4\n",
      "Predicting variants for PAS = BRCA1.5\n",
      "Predicting variants for PAS = BRCA1.6\n",
      "Predicting variants for PAS = BRCA1.7\n",
      "Predicting variants for PAS = BRCA1.8\n",
      "Predicting variants for PAS = BRCA1.9\n",
      "Predicting variants for PAS = BRCA1.10\n",
      "Predicting variants for PAS = BRCA2.1\n",
      "Predicting variants for PAS = BRCA2.2\n",
      "Predicting variants for PAS = BRCA2.3\n",
      "Predicting variants for PAS = BRCA2.4\n",
      "Predicting variants for PAS = BRCA2.5\n",
      "Predicting variants for PAS = BRCA2.6\n",
      "Predicting variants for PAS = BRCA2.7\n",
      "Predicting variants for PAS = BRCA2.8\n",
      "Predicting variants for PAS = BRCA2.9\n",
      "Predicting variants for PAS = CEBPA.1\n",
      "Predicting variants for PAS = CEBPA.2\n",
      "Predicting variants for PAS = CEBPA.3\n",
      "Predicting variants for PAS = CEBPA.4\n",
      "Predicting variants for PAS = CEBPA.5\n",
      "Predicting variants for PAS = CEBPA.6\n",
      "Predicting variants for PAS = CEBPA.7\n",
      "Predicting variants for PAS = CEBPA.8\n",
      "Predicting variants for PAS = CEBPA.9\n",
      "Predicting variants for PAS = CEBPA.10\n",
      "Predicting variants for PAS = CEBPA.11\n",
      "Predicting variants for PAS = CEBPA.12\n",
      "Predicting variants for PAS = DDX41.1\n",
      "Predicting variants for PAS = DDX41.2\n",
      "Predicting variants for PAS = DDX41.3\n",
      "Predicting variants for PAS = DDX41.4\n",
      "Predicting variants for PAS = DDX41.5\n",
      "Predicting variants for PAS = ETV6.1\n",
      "Predicting variants for PAS = ETV6.2\n",
      "Predicting variants for PAS = ETV6.3\n",
      "Predicting variants for PAS = ETV6.4\n",
      "Predicting variants for PAS = ETV6.5\n",
      "Predicting variants for PAS = ETV6.6\n",
      "Predicting variants for PAS = ETV6.7\n",
      "Predicting variants for PAS = ETV6.8\n",
      "Predicting variants for PAS = GATA2.1\n",
      "Predicting variants for PAS = GATA2.2\n",
      "Predicting variants for PAS = GATA2.3\n",
      "Predicting variants for PAS = GATA2.4\n",
      "Predicting variants for PAS = GATA2.5\n",
      "Predicting variants for PAS = GATA2.6\n",
      "Predicting variants for PAS = GATA2.7\n",
      "Predicting variants for PAS = GATA2.8\n",
      "Predicting variants for PAS = GATA2.9\n",
      "Predicting variants for PAS = GATA2.10\n",
      "Predicting variants for PAS = GATA2.11\n",
      "Predicting variants for PAS = GATA2.12\n",
      "Predicting variants for PAS = GATA2.13\n",
      "Predicting variants for PAS = GATA2.14\n",
      "Predicting variants for PAS = GATA2.15\n",
      "Predicting variants for PAS = MLH1.1\n",
      "Predicting variants for PAS = MLH1.2\n",
      "Predicting variants for PAS = MLH1.3\n",
      "Predicting variants for PAS = MLH1.4\n",
      "Predicting variants for PAS = MLH1.5\n",
      "Predicting variants for PAS = MLH1.6\n",
      "Predicting variants for PAS = MLH1.7\n",
      "Predicting variants for PAS = MLH1.8\n",
      "Predicting variants for PAS = MLH1.9\n",
      "Predicting variants for PAS = MLH1.10\n",
      "Predicting variants for PAS = MLH1.11\n",
      "Predicting variants for PAS = MLH1.12\n",
      "Predicting variants for PAS = MLH1.13\n",
      "Predicting variants for PAS = MLH1.14\n",
      "Predicting variants for PAS = MLH1.15\n",
      "Predicting variants for PAS = MLH1.16\n",
      "Predicting variants for PAS = MLH1.17\n",
      "Predicting variants for PAS = MSH2.1\n",
      "Predicting variants for PAS = MSH2.2\n",
      "Predicting variants for PAS = MSH2.3\n",
      "Predicting variants for PAS = MSH2.4\n",
      "Predicting variants for PAS = MSH2.5\n",
      "Predicting variants for PAS = MSH2.6\n",
      "Predicting variants for PAS = MSH2.7\n",
      "Predicting variants for PAS = MSH2.8\n",
      "Predicting variants for PAS = MSH2.9\n",
      "Predicting variants for PAS = MSH2.10\n",
      "Predicting variants for PAS = MSH2.11\n",
      "Predicting variants for PAS = MSH2.12\n",
      "Predicting variants for PAS = MSH2.13\n",
      "Predicting variants for PAS = MSH2.14\n",
      "Predicting variants for PAS = MSH2.15\n",
      "Predicting variants for PAS = MSH2.16\n",
      "Predicting variants for PAS = MSH2.17\n",
      "Predicting variants for PAS = MSH2.18\n",
      "Predicting variants for PAS = MSH2.19\n",
      "Predicting variants for PAS = MSH2.20\n",
      "Predicting variants for PAS = MSH2.21\n",
      "Predicting variants for PAS = MSH2.22\n",
      "Predicting variants for PAS = MSH2.23\n",
      "Predicting variants for PAS = MSH2.24\n",
      "Predicting variants for PAS = MSH6.1\n",
      "Predicting variants for PAS = MSH6.2\n",
      "Predicting variants for PAS = MSH6.3\n",
      "Predicting variants for PAS = MSH6.4\n",
      "Predicting variants for PAS = MSH6.5\n",
      "Predicting variants for PAS = MSH6.6\n",
      "Predicting variants for PAS = MSH6.7\n",
      "Predicting variants for PAS = MSH6.8\n",
      "Predicting variants for PAS = MSH6.9\n",
      "Predicting variants for PAS = MSH6.10\n",
      "Predicting variants for PAS = MSH6.11\n",
      "Predicting variants for PAS = MSH6.12\n",
      "Predicting variants for PAS = MSH6.13\n",
      "Predicting variants for PAS = MSH6.14\n",
      "Predicting variants for PAS = MSH6.15\n",
      "Predicting variants for PAS = MUTYH.1\n",
      "Predicting variants for PAS = MUTYH.2\n",
      "Predicting variants for PAS = MUTYH.3\n",
      "Predicting variants for PAS = MUTYH.4\n",
      "Predicting variants for PAS = MUTYH.5\n",
      "Predicting variants for PAS = MUTYH.6\n",
      "Predicting variants for PAS = PALB2.1\n",
      "Predicting variants for PAS = PALB2.2\n",
      "Predicting variants for PAS = PALB2.3\n",
      "Predicting variants for PAS = PALB2.4\n",
      "Predicting variants for PAS = PALB2.5\n",
      "Predicting variants for PAS = PMS2.1\n",
      "Predicting variants for PAS = PMS2.2\n",
      "Predicting variants for PAS = PMS2.3\n",
      "Predicting variants for PAS = PMS2.4\n",
      "Predicting variants for PAS = PMS2.5\n",
      "Predicting variants for PAS = PMS2.6\n",
      "Predicting variants for PAS = PTEN.1\n",
      "Predicting variants for PAS = PTEN.2\n",
      "Predicting variants for PAS = PTEN.3\n",
      "Predicting variants for PAS = PTEN.4\n",
      "Predicting variants for PAS = PTEN.5\n",
      "Predicting variants for PAS = PTEN.6\n",
      "Predicting variants for PAS = PTEN.7\n",
      "Predicting variants for PAS = PTEN.8\n",
      "Predicting variants for PAS = PTEN.9\n",
      "Predicting variants for PAS = PTEN.10\n",
      "Predicting variants for PAS = PTEN.11\n",
      "Predicting variants for PAS = PTEN.12\n",
      "Predicting variants for PAS = PTEN.13\n",
      "Predicting variants for PAS = PTEN.14\n",
      "Predicting variants for PAS = PTEN.15\n",
      "Predicting variants for PAS = PTEN.16\n",
      "Predicting variants for PAS = PTEN.17\n",
      "Predicting variants for PAS = PTEN.18\n",
      "Predicting variants for PAS = PTEN.19\n",
      "Predicting variants for PAS = PTEN.20\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Predicting variants for PAS = PTEN.21\n",
      "Predicting variants for PAS = PTEN.22\n",
      "Predicting variants for PAS = PTEN.23\n",
      "Predicting variants for PAS = PTEN.24\n",
      "Predicting variants for PAS = PTEN.25\n",
      "Predicting variants for PAS = PTEN.26\n",
      "Predicting variants for PAS = PTEN.27\n",
      "Predicting variants for PAS = PTEN.28\n",
      "Predicting variants for PAS = PTEN.29\n",
      "Predicting variants for PAS = PTEN.30\n",
      "Predicting variants for PAS = PTEN.31\n",
      "Predicting variants for PAS = PTEN.32\n",
      "Predicting variants for PAS = PTEN.33\n",
      "Predicting variants for PAS = PTEN.34\n",
      "Predicting variants for PAS = PTEN.35\n",
      "Predicting variants for PAS = PTEN.36\n",
      "Predicting variants for PAS = PTEN.37\n",
      "Predicting variants for PAS = PTEN.38\n",
      "Predicting variants for PAS = PTEN.39\n",
      "Predicting variants for PAS = PTEN.40\n",
      "Predicting variants for PAS = PTEN.41\n",
      "Predicting variants for PAS = PTEN.42\n",
      "Predicting variants for PAS = PTEN.43\n",
      "Predicting variants for PAS = PTEN.44\n",
      "Predicting variants for PAS = PTEN.45\n",
      "Predicting variants for PAS = PTEN.46\n",
      "Predicting variants for PAS = PTEN.47\n",
      "Predicting variants for PAS = RUNX1.1\n",
      "Predicting variants for PAS = RUNX1.2\n",
      "Predicting variants for PAS = RUNX1.3\n",
      "Predicting variants for PAS = RUNX1.4\n",
      "Predicting variants for PAS = RUNX1.5\n",
      "Predicting variants for PAS = RUNX1.6\n",
      "Predicting variants for PAS = RUNX1.7\n",
      "Predicting variants for PAS = RUNX1.8\n",
      "Predicting variants for PAS = RUNX1.9\n",
      "Predicting variants for PAS = RUNX1.10\n",
      "Predicting variants for PAS = RUNX1.11\n",
      "Predicting variants for PAS = RUNX1.12\n",
      "Predicting variants for PAS = RUNX1.13\n",
      "Predicting variants for PAS = RUNX1.14\n",
      "Predicting variants for PAS = RUNX1.15\n",
      "Predicting variants for PAS = RUNX1.16\n",
      "Predicting variants for PAS = RUNX1.17\n",
      "Predicting variants for PAS = RUNX1.18\n",
      "Predicting variants for PAS = RUNX1.19\n",
      "Predicting variants for PAS = RUNX1.20\n",
      "Predicting variants for PAS = RUNX1.21\n",
      "Predicting variants for PAS = RUNX1.22\n",
      "Predicting variants for PAS = RUNX1.23\n",
      "Predicting variants for PAS = RUNX1.24\n",
      "Predicting variants for PAS = RUNX1.25\n",
      "Predicting variants for PAS = RUNX1.26\n",
      "Predicting variants for PAS = RUNX1.27\n",
      "Predicting variants for PAS = RUNX1.28\n",
      "Predicting variants for PAS = RUNX1.29\n",
      "Predicting variants for PAS = RUNX1.30\n",
      "Predicting variants for PAS = RUNX1.31\n",
      "Predicting variants for PAS = RUNX1.32\n",
      "Predicting variants for PAS = RUNX1.33\n",
      "Predicting variants for PAS = RUNX1.34\n",
      "Predicting variants for PAS = TP53.1\n",
      "Predicting variants for PAS = TP53.2\n",
      "Predicting variants for PAS = TP53.3\n",
      "Predicting variants for PAS = TP53.4\n",
      "Predicting variants for PAS = TP53.5\n",
      "Predicting variants for PAS = TP53.6\n",
      "Predicting variants for PAS = TP53.7\n",
      "Predicting variants for PAS = TP53.8\n",
      "Predicting variants for PAS = TP53.9\n",
      "Predicting variants for PAS = TP53.10\n",
      "Predicting variants for PAS = TP53.11\n",
      "Predicting variants for PAS = TP53.12\n",
      "Predicting variants for PAS = TP53.13\n",
      "Predicting variants for PAS = TP53.14\n",
      "Predicting variants for PAS = TP53.15\n",
      "Predicting variants for PAS = TP53.16\n",
      "Predicting variants for PAS = TP53.17\n"
     ]
    }
   ],
   "source": [
    "#Perform in-silico saturation mutagenesis\n",
    "\n",
    "aparent_encoder = get_aparent_encoder()\n",
    "\n",
    "isoform_start = 77\n",
    "isoform_end = 77 + 40\n",
    "\n",
    "variant_dict = {\n",
    "    'gene' : [],\n",
    "    'gene_id' : [],\n",
    "    #'ref_seq' : [],\n",
    "    'chrom' : [],\n",
    "    'strand' : [],\n",
    "    'site_type' : [],\n",
    "    'native_usage' : [],\n",
    "    #'var_seq' : [],\n",
    "    'var_position' : [],\n",
    "    'ref_nucleotide' : [],\n",
    "    'var_nucleotide' : [],\n",
    "    'delta_logodds_fixed_window' : []\n",
    "}\n",
    "\n",
    "for _, row in polyadb_df.iterrows() :\n",
    "    gene = row['gene']\n",
    "    gene_id = row['gene_id']\n",
    "    \n",
    "    print(\"Predicting variants for PAS = \" + str(gene_id))\n",
    "    \n",
    "    ref_seq = row['wide_seq_ext'][175-70:175-70+205]\n",
    "    \n",
    "    chrom = row['chrom']\n",
    "    strand = row['strand']\n",
    "    site_type = row['site_type']\n",
    "    ref_usage = row['ratio']\n",
    "    \n",
    "    pas_pos = row['pas_pos']\n",
    "    \n",
    "    seq_start = 0\n",
    "    if strand == '+' :\n",
    "        #seq_start = pas_pos - 70\n",
    "        seq_start = pas_pos - 70 + 1\n",
    "    else :\n",
    "        #seq_start = pas_pos - (205 - 76)\n",
    "        seq_start = pas_pos - (205 - 70)\n",
    "    \n",
    "    seq_end = seq_start + 205\n",
    "    \n",
    "    #Predict reference sequence with APARENT model\n",
    "    _, ref_cut_pred = aparent_model.predict(x=aparent_encoder([ref_seq]))\n",
    "    \n",
    "    ref_iso_pred_fixed_window = np.sum(ref_cut_pred[0, isoform_start:isoform_end])\n",
    "    \n",
    "    #Predict all variants\n",
    "    for pos in range(205) :\n",
    "        for base in ['A', 'C', 'G', 'T'] :\n",
    "            var_seq = ref_seq[:pos] + base + ref_seq[pos+1:]\n",
    "            if var_seq == ref_seq :\n",
    "                continue\n",
    "            \n",
    "            #Predict variant sequence with APARENT model\n",
    "            _, var_cut_pred = aparent_model.predict(x=aparent_encoder([var_seq]))\n",
    "            \n",
    "            var_iso_pred_fixed_window = np.sum(var_cut_pred[0, isoform_start:isoform_end])\n",
    "            \n",
    "            delta_logodds_fixed_window = np.log(var_iso_pred_fixed_window / (1. - var_iso_pred_fixed_window)) - np.log(ref_iso_pred_fixed_window / (1. - ref_iso_pred_fixed_window))\n",
    "            \n",
    "            \n",
    "            var_position = 0\n",
    "            var_nucleotide = 'A'\n",
    "            if strand == '+' :\n",
    "                var_position = seq_start + pos\n",
    "                var_nucleotide = base\n",
    "            else :\n",
    "                var_position = seq_end - pos\n",
    "                if base == 'A' :\n",
    "                    var_nucleotide = 'T'\n",
    "                elif base == 'C' :\n",
    "                    var_nucleotide = 'G'\n",
    "                elif base == 'G' :\n",
    "                    var_nucleotide = 'C'\n",
    "                elif base == 'T' :\n",
    "                    var_nucleotide = 'A'\n",
    "            \n",
    "            ref_nucleotide = 'A'\n",
    "            if strand == '+' :\n",
    "                ref_nucleotide = ref_seq[pos]\n",
    "            else :\n",
    "                if ref_seq[pos] == 'A' :\n",
    "                    ref_nucleotide = 'T'\n",
    "                elif ref_seq[pos] == 'C' :\n",
    "                    ref_nucleotide = 'G'\n",
    "                elif ref_seq[pos] == 'G' :\n",
    "                    ref_nucleotide = 'C'\n",
    "                elif ref_seq[pos] == 'T' :\n",
    "                    ref_nucleotide = 'A'\n",
    "            \n",
    "            variant_dict['gene'].append(gene)\n",
    "            variant_dict['gene_id'].append(gene_id)\n",
    "            #variant_dict['ref_seq'].append(ref_seq)\n",
    "            variant_dict['chrom'].append(chrom)\n",
    "            variant_dict['strand'].append(strand)\n",
    "            variant_dict['site_type'].append(site_type)\n",
    "            variant_dict['native_usage'].append(ref_usage)\n",
    "            #variant_dict['var_seq'].append(var_seq)\n",
    "            variant_dict['var_position'].append(var_position)\n",
    "            variant_dict['ref_nucleotide'].append(ref_nucleotide)\n",
    "            variant_dict['var_nucleotide'].append(var_nucleotide)\n",
    "            variant_dict['delta_logodds_fixed_window'].append(delta_logodds_fixed_window)\n",
    "\n",
    "variant_df = pd.DataFrame(variant_dict)\n",
    "#variant_df = variant_df[['gene','gene_id','ref_seq','chrom','strand','site_type','native_usage','var_seq','var_position','ref_nucleotide','var_nucleotide','delta_logodds_fixed_window']]\n",
    "variant_df = variant_df[['gene','gene_id','chrom','strand','site_type','native_usage','var_position','ref_nucleotide','var_nucleotide','delta_logodds_fixed_window']]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "variant_df['native_usage_pred_var'] = 1. - 1. / (1. + (variant_df['native_usage'] / (1. - variant_df['native_usage'])) * np.exp(variant_df['delta_logodds_fixed_window']))\n",
    "variant_df['delta_usage'] = variant_df['native_usage_pred_var'] - variant_df['native_usage']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Store variant prediction dataframe\n",
    "\n",
    "variant_df.to_csv('variant_predictions_polyadb_no_sequences.csv', sep='\\t')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/johli/anaconda3/envs/aparent/lib/python3.6/site-packages/IPython/core/interactiveshell.py:3049: DtypeWarning: Columns (18) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    }
   ],
   "source": [
    "#Load ClinVar variant summary\n",
    "\n",
    "clinvar_df = pd.read_csv('clinvar_variant_summary_20200205.txt', sep='\\t')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "clinvar_df = clinvar_df.query(\"Type == 'single nucleotide variant' and Assembly == 'GRCh37'\").reset_index(drop=True).copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "variant_df['flat_id'] = variant_df['chrom'].apply(str) + \"_\" + variant_df['var_position'].apply(str) + \"_\" + variant_df['var_nucleotide'].apply(str)\n",
    "clinvar_df['flat_id'] = 'chr' + clinvar_df['Chromosome'].apply(str) + \"_\" + clinvar_df['Start'].apply(str) + \"_\" + clinvar_df['AlternateAllele'].apply(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "clinvar_df = clinvar_df[['flat_id', 'Name', 'GeneSymbol', 'Chromosome', 'Start', 'Stop', 'ReferenceAllele', 'AlternateAllele', 'ClinicalSignificance']]\n",
    "\n",
    "clinvar_df = clinvar_df.rename(columns={\n",
    "    'Name' : 'clinvar_id',\n",
    "    'GeneSymbol' : 'clinvar_gene',\n",
    "    'Chromosome' : 'clinvar_chrom',\n",
    "    'Start' : 'clinvar_start',\n",
    "    'End' : 'clinvar_end',\n",
    "    'ReferenceAllele' : 'clinvar_ref_nucleotide',\n",
    "    'AlternateAllele' : 'clinvar_var_nucleotide',\n",
    "    'ClinicalSignificance' : 'clinvar_significance'\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "mapped_variant_df = variant_df.join(clinvar_df.set_index('flat_id'), on='flat_id', how='left')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "mapped_variant_df = mapped_variant_df.rename(columns={\n",
    "    'gene_id' : 'pas_id',\n",
    "    'var_position' : 'position',\n",
    "    'ref_nucleotide' : 'ref_allele',\n",
    "    'var_nucleotide' : 'var_allele',\n",
    "    'delta_logodds_fixed_window' : 'delta_logodds'\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "#mapped_variant_df = mapped_variant_df[['gene', 'chrom', 'strand', 'pas_id', 'site_type', 'ref_seq', 'var_seq', 'position', 'ref_allele', 'var_allele', 'delta_logodds', 'native_usage', 'delta_usage', 'clinvar_id', 'clinvar_significance']]\n",
    "mapped_variant_df = mapped_variant_df[['gene', 'chrom', 'strand', 'pas_id', 'site_type', 'position', 'ref_allele', 'var_allele', 'delta_logodds', 'native_usage', 'delta_usage', 'clinvar_id', 'clinvar_significance']]\n",
    "\n",
    "mapped_variant_df.to_csv(\"all_variant_predictions_polyadb_no_sequences.csv\", sep='\\t', index=False)\n",
    "\n",
    "mapped_variant_df.loc[~mapped_variant_df['clinvar_id'].isnull()].to_csv(\"clinvar_variant_predictions_polyadb_no_sequences.csv\", sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "mapped_variant_df = mapped_variant_df.sort_values(by='delta_logodds').copy().reset_index(drop=True)\n",
    "\n",
    "mapped_variant_df.to_csv(\"all_variant_predictions_polyadb_no_sequences_sorted.csv\", sep='\\t', index=False)\n",
    "\n",
    "mapped_variant_df.loc[~mapped_variant_df['clinvar_id'].isnull()].to_csv(\"clinvar_variant_predictions_polyadb_no_sequences_sorted.csv\", sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:aparent]",
   "language": "python",
   "name": "conda-env-aparent-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
