<!doctype html>
<html lang="en">
  <head>
    <!-- Required meta tags -->
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">

    <!-- Bootstrap CSS -->
    <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.3.1/css/bootstrap.min.css" integrity="sha384-ggOyR0iXCbMQv3Xipma34MD+dH/1fQ784/j6cY/iJTQUOhcWr7x9JvoRxT2MZw1T" crossorigin="anonymous">

    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap-select@1.13.7/dist/css/bootstrap-select.min.css">

    <title>APARENT</title>
  </head>
  <body>
    

    


  <div class="d-flex flex-column flex-md-row align-items-center p-3 px-md-4 mb-3 bg-white border-bottom shadow-sm">
    <h5 class="my-0 mr-md-auto font-weight-normal">APARENT</h5>
    <nav class="my-2 my-md-0 mr-md-3">
      <a class="p-2 text-dark" href="https://apa.cs.washington.edu">Home</a>
      <a class="p-2 text-dark" href="https://apa.cs.washington.edu/mutagenesis">Mutagenesis</a>
      <a class="p-2 text-dark" href="https://apa.cs.washington.edu/apa">APA Isoforms</a>
      <a class="p-2 text-dark" href="https://apa.cs.washington.edu/pas">PAS Score</a>
      <a class="p-2 text-dark" href="https://apa.cs.washington.edu/variant">Variants</a>
      <a class="p-2 text-dark" href="https://apa.cs.washington.edu/detect">PolyA Detection</a>
    </nav>
    <a class="btn btn-outline-primary" href="https://github.com/johli/aparent">GitHub</a>
  </div>



  <div class="px-3 py-3 pt-md-5 pb-md-4 mx-auto text-center"> <!-- pricing-header -->
    <div class='container'>
      <div class='row justify-content-center'>
        <div class="col-sm-8 col-md-8 col-lg-8 col-xl-8">
          <img src="https://apa.cs.washington.edu/img/aparent_logo_cropped.jpg" class="img-fluid" alt="APA Logo">
        </div>
      </div>
      </br>
      <div class='row justify-content-center'>
        <div class="col-sm-10 col-md-10 col-lg-10 col-xl-10">
          <p class="lead">APARENT is a deep neural network that can predict human 3' UTR Alternative Polyadenylation (APA) and the effects of genetic variants on APA. The network was trained on >3.5 million randomized 3' UTR poly-A signals expressed on mini gene reporters in HEK293.
          <br/>
          <br/>
          APARENT was described in <a href='https://doi.org/10.1016/j.cell.2019.04.046'>Bogard et al, Cell 2019 in press.</a><br/>
          It was developed in <a href='https://www.seeliglab.org/'>Seelig Lab</a> at the University of Washington.<br/>
          <br/>
          Contact jlinder2 (at) cs.washington.edu for any questions about the model or data.
          </p>
        </div>
      </div>
    </div>
    <br/>
  </div>

  <div class="container">
    <div class="card-deck mb-3 text-center">
      <div class="card border-primary mb-4 shadow-sm">
        <div class="card-header">
          <h4 class="my-0 font-weight-normal">Mutagenesis</h4>
        </div>
        <div class="card-body">
          <img src="https://apa.cs.washington.edu/img/mutagenesis_logo.png" class="img-fluid" alt="Mutagenesis">
          <ul class="list-unstyled mt-3 mb-4">
            <li>1. Choose a PAS from the genome</li>
            <li>2. Predict all possible SNVs</li>
            <li>3. Inspect Cleavage alteration</li>
            <li>4. Inspect Isoform fold change</li>
          </ul>
          <a class="btn btn-lg btn-block btn-outline-primary" href="https://apa.cs.washington.edu/mutagenesis">Go</a>
        </div>
      </div>
      <div class="card border-primary mb-4 shadow-sm">
        <div class="card-header">
          <h4 class="my-0 font-weight-normal">APA Isoforms</h4>
        </div>
        <div class="card-body">
          <img src="https://apa.cs.washington.edu/img/native_apa_logo.png" class="img-fluid" alt="Mutagenesis">
          <ul class="list-unstyled mt-3 mb-4">
            <li>1. Choose APA sites from the genome</li>
            <li>2. Specifcy Isoforms</li>
            <li>3. Predict Isoform Abundance</li>
            <li>4. (Optionally) Alter sequences</li>
          </ul>
          <a class="btn btn-lg btn-block btn-outline-primary" href="https://apa.cs.washington.edu/apa">Go</a>
        </div>
      </div>
      <div class="card border-primary mb-4 shadow-sm">
        <div class="card-header">
          <h4 class="my-0 font-weight-normal">PAS Score</h4>
        </div>
        <div class="card-body">
          <img src="https://apa.cs.washington.edu/img/pas_score_logo.png" class="img-fluid" alt="Mutagenesis">
          <ul class="list-unstyled mt-3 mb-4">
            <li>1. Choose a PAS from the genome</li>
            <li>2. Specifcy Isoform</li>
            <li>3. Predict PAS Score (strength)</li>
            <li>4. (Optionally) Alter sequence</li>
          </ul>
          <a class="btn btn-lg btn-block btn-outline-primary" href="https://apa.cs.washington.edu/pas">Go</a>
        </div>
      </div>
    </div>
  </div>

  <div class="container">
    <div class="card-deck mb-3 text-center">
      <div class="card border-primary mb-4 shadow-sm">
        <div class="card-header">
          <h4 class="my-0 font-weight-normal">Variants</h4>
        </div>
        <div class="card-body">
          <img src="https://apa.cs.washington.edu/img/variant_logo.png" class="img-fluid" alt="Variants">
          <ul class="list-unstyled mt-3 mb-4">
            <li>1. Choose a PAS from the genome</li>
            <li>2. Specify the variant sequence</li>
            <li>3. Inspect Cleavage alteration</li>
            <li>4. Inspect Isoform fold change</li>
          </ul>
          <a class="btn btn-lg btn-block btn-outline-primary" href="https://apa.cs.washington.edu/variant">Go</a>
        </div>
      </div>
      <div class="card border-primary mb-4 shadow-sm">
        <div class="card-header">
          <h4 class="my-0 font-weight-normal">PolyA Detection</h4>
        </div>
        <div class="card-body">
          <img src="https://apa.cs.washington.edu/img/detect_logo.png" class="img-fluid" alt="Detection">
          <ul class="list-unstyled mt-3 mb-4">
            <li>1. Choose pA sites from the genome</li>
            <li>2. Or specify a custom sequence</li>
            <li>3. Inspect PolyA profile</li>
            <li>4. Zoom in and detect PASs</li>
          </ul>
          <a class="btn btn-lg btn-block btn-outline-primary" href="https://apa.cs.washington.edu/detect">Go</a>
        </div>
      </div>
      <div class="card border-light mb-4 shadow-sm">
        <div class="card-header">
          <h4 class="my-0 font-weight-normal">GitHub</h4>
        </div>
        <div class="card-body">
          <img src="https://apa.cs.washington.edu/img/gitgo_logo.png" class="img-fluid" alt="GitHub">
          <ul class="list-unstyled mt-3 mb-4">
            <li>- Run APARENT in batch mode</li>
            <li>- Optimize sequences with SeqProp</li>
            <li>- Visualize network features</li>
            <li>- Integrate APARENT in your system</li>
          </ul>
          <a class="btn btn-lg btn-block btn-outline-success" href="https://github.com/johli/aparent">GitHub</a>
        </div>
      </div>
    </div>
  </div>

  <footer class="my-5 pt-5 text-muted text-center text-small">
    <p class="mb-1">Seelig Lab</p>
    <a href="https://www.seeliglab.org/">Website</a> | jlinder2 (at) cs.washington.edu
  </footer>
    

  </body>
</html>
